### PR TITLE
AppendFeaturesFromCallable InputTransform

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

This PR adds the functionality of transfer learning and feature engineering within a model via a new input transform called `AppendFeaturesFromCallable` as discussed in https://github.com/pytorch/botorch/discussions/1307. This implementation does not alter the original `AppendFeatures` transform but adds a new one to keep things clean and separate.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

### CODE REVIEW

File: test/models/transforms/test_input.py [hunk 1]
```diff
@@ -847,6 +848,88 @@ def test_w_skip_expand(self):
...
+class TestAppendFeaturesFromCallable(BotorchTestCase):
+    def test_append_features_from_callable(self):
+        # basic init, provided indices
+        for dtype in [torch.float, torch.double]:
+            tkwargs = {"device": self.device, "dtype": dtype}
+            with self.assertRaises(ValueError):
+                transform = AppendFeaturesFromCallable(
+                    d=2,
+                    indices=[0, 1, 2],
+                    f=lambda x: torch.sum(x, dim=-1, keepdim=True),
```
Comment (Balandat) [inline test/models/transforms/test_input.py:860]: This is not wrong, but let's use the defined order to make it consistent / more readable
```suggestion
                    f=lambda x: torch.sum(x, dim=-1, keepdim=True),
                    d=2,
                    indices=[0, 1, 2],  
```

File: test/models/transforms/test_input.py [hunk 2]
```diff
@@ -847,6 +848,88 @@ def test_w_skip_expand(self):
...
+            with self.assertRaises(ValueError):
+                transform = AppendFeaturesFromCallable(
+                    d=2, indices=[0, 0], f=lambda x: torch.sum(x, dim=-1, keepdim=True)
+                )
+            with self.assertRaises(ValueError):
+                transform = AppendFeaturesFromCallable(
+                    d=2, indices=[], f=lambda x: torch.sum(x, dim=-1, keepdim=True)
+                )
+        # test functionality
+        X = torch.rand(10, 3).to(**tkwargs)
```
Comment (Balandat) [inline test/models/transforms/test_input.py:875]:
```suggestion
        X = torch.rand(10, 3, **tkwargs)
```

File: test/models/transforms/test_input.py [hunk 3]
```diff
@@ -847,6 +848,88 @@ def test_w_skip_expand(self):
...
+        transform = AppendFeaturesFromCallable(
+            d=3,
+            f=lambda x: torch.sum(x, dim=-1, keepdim=True),
```
Comment (Balandat) [inline test/models/transforms/test_input.py:878]:
```suggestion
            f=lambda x: torch.sum(x, dim=-1, keepdim=True),
            d=3,
```

File: test/models/transforms/test_input.py (hunk 1 above)
Comment (Balandat) [reply test/models/transforms/test_input.py:860]: How about defining `lambda x: torch.sum(x, dim=-1, keepdim=True)` as a function in the test and using that (rather than re-defining the lambda function in every call)?

File: test/models/transforms/test_input.py (hunk 3 above)
Comment (Balandat) [reply test/models/transforms/test_input.py:878]: also in a couple of other places below

**Review by Balandat — APPROVED** [review]
Overall this lgtm - a few nits other than that this is good to go.

File: botorch/models/transforms/input.py [hunk 4]
```diff
@@ -1030,6 +1030,78 @@ def transform(self, X: Tensor) -> Tensor:
...
+        if (indices is not None) and (len(indices) == 0):
+            raise ValueError("`indices` list is empty!")
+        if (indices is not None) and (len(indices) > 0):
+            indices = torch.tensor(indices, dtype=torch.long)
+            if len(indices) > d:
+                raise ValueError("Can provide at most `d` indices!")
+            if (indices > d - 1).any():
+                raise ValueError("Elements of `indices` have to be smaller than `d`!")
+            if len(indices.unique()) != len(indices):
+                raise ValueError("Elements of `indices` tensor must be unique!")
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py:1076]: Good idea, I just added a call to `normalize_indices`

File: botorch/models/transforms/input.py [hunk 5]
```diff
@@ -1030,6 +1030,78 @@ def transform(self, X: Tensor) -> Tensor:
...
+
+        Args:
+            X: A `batch_shape x q x d`-dim tensor of inputs.
+
+        Returns:
+            A `batch_shape x q x (d + d_f)`-dim tensor of appended inputs.
+        """
+
+        features = self._f(X[..., self.indices])
+        expanded_X = X.unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py:1099]: I oriented myself on the implementation of `AppendFeatures`, but fooled myself when replacing the final `view` with a `squeeze`. You are correct having `q=1` removes the `q`-dimension. I fixed it and added respective test cases ;)

File: test/models/transforms/test_input.py [hunk 6]
```diff
@@ -847,6 +848,88 @@ def test_w_skip_expand(self):
...
+        )
+        X_transformed = transform(X)
+        assert X_transformed.shape == torch.Size((10, 4))
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply test/models/transforms/test_input.py]: Done!

File: test/models/transforms/test_input.py (hunk 1 above)
Comment (jduerholt) [reply test/models/transforms/test_input.py]: Done ;)

**Comment by saitcakmak** [comment]
Hi @jduerholt. Thanks for putting this together! Given that this shares a lot of functionality with the existing `AppendFeatures` transform, I think it might be better if we merged the two together into a single transform that supports both fixed set of features and a callable.

**Comment by jduerholt** [comment]
Hi @saitcakmak, ok, let's do it. You suggested it already in the beginning, but I was not quite sure how to integrateit  in the best way with the expansion along the `q`-dimension in `AppendFeatures`, as this is not needed in our use case. For this reason, I implemented it in a standalone transform. But I am absolutely open to unite it with the orginal `AppendFeatures`. 

In case of providing a callable, should we then only allow for appending along the last dimension or also the `q`-dimension? 

In principle like shown in the snippet below: Or do you want to have the functionality to do the same as in `AppendFeatures` via a callable, namely also the expansion around the `q`-dimension?

```python

def transform(self, X):
    if self._f:
        ### do stuff from AppendFeaturesFromCallable
    else:
        ### do orginal AppendFeatures Stuff
    return appended_X.view(*X.shape[:-2], -1, appended_X.shape[-1])
```

Should we first merge this PR, and then creating a new PR for the combination or first unite it and then integrate it into main?

**Comment by jduerholt** [comment]
From my perspective, the main difference is currently that the callable always produces a `batch_shape x n x d_f` tensor wheres `AppendFeatures` works with a `n_f x d_f` tensor in which the batch dimension is ignored. Furthermore we only append to the last dimension. We would somehow need a flag to tell the `InputTransform` in which mode it should operate and how to interprete the return tensor of the callable. What do you think?

**Comment by saitcakmak** [comment]
> In case of providing a callable, should we then only allow for appending along the last dimension or also the q-dimension?

I'd say let's support both. If the callable returns `q=1`, then the q-batch dimension would remain the same (the current setup). If it returns `q>1`, then it would get expanded as is done in `AppendFeatures`. `InputPerturbation` transform has an example of how this could be done. Something like
```
if self.feature_set is not None:
    features = self.feature_set.expand(...)  # to match X batch shape
else:
    features = self._f(X)
expanded_X = X.unsqueeze(dim=-2).expand(*X.shape[:-1], features.shape[-2], -1)
appended_X = torch.cat([expanded_X, features], dim=-1)
return appended_X.view(*X.shape[:2], -1, X.shape[-1])
```

> Should we first merge this PR, and then creating a new PR for the combination or first unite it and then integrate it into main?

Let's iterate on this one if you don't mind.

**Comment by jduerholt** [comment]
No problem, we can intereate on this one. The thing which I currently do not understand is your distiction on the `q`-dimension. Also in my current case the `q`-dim could be >1 but still we only want to expand it in the features dimension. This can be seen in line 900 of `test_input.py` of this PR. There we have a callable called `f2` that appends always two features. In practice, this could be a multi-output model used in the transfer learning.

**Comment by jduerholt** [comment]
Another question: Would you then also integrate the `indices` keyword in the transform or would you see this filtering as part of the callable?

**Comment by saitcakmak** [comment]
Sorry about the confusion there. In the current setup, if `X` is `batch x q x d`, `f` returns `batch x q x d_f`. To support expanding the `q` dimension, we will modify this slightly. `f` will return `batch x q x n_f x d_f`. If `n_f = 1` (what I referred to as `q` before), then the `appended_X` will be `batch x q x (d + d_f)`. If `n_f > 1`, it will be `batch x (q x n_f) x (d + d_f)`.

**Comment by saitcakmak** [comment]
> Would you then also integrate the indices keyword in the transform or would you see this filtering as part of the callable?

Your call. I don't see any harm in keeping it. If we keep it, let's throw an error if it is specified when not using `f`.

**Comment by jduerholt** [comment]
Perfect, now I understand it ;) I will try to refine it this week.

**Comment by jduerholt** [comment]
@saitcakmak I tried to unite both functionalities in the new commit. Just have a look, if this fits to your idea. So far, `skip_expand` is not supported in case of providing a callable.

**Review by saitcakmak — COMMENTED** [review]
Thanks for the changes. A few minor comments in line, otherwise this looks great.

Re `skip_expand`: Is there any blocker to supporting this? It seems to me that the code should work as is as long as `_f` expects an already expanded `X` (which is for the user to make sure). It should be ok either way.

File: botorch/models/transforms/input.py [hunk 7]
```diff
@@ -987,16 +1011,49 @@ def __init__(
...
+            if feature_set.dim() != 2:
+                raise ValueError("`feature_set` must be an `n_f x d_f`-dim tensor!")
+            self.register_buffer("feature_set", feature_set)
+            self._use_callable = False
+        if f is not None:
+            if skip_expand:
+                raise ValueError(
+                    "`skip_expand` option is not supported in case of using a callable"
+                )
+            if d is None:
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py]: It was for safety purposes to prevent the user to give indices twice, but it makes the whole thing unhandy. I remove it ;)

File: botorch/models/transforms/input.py [hunk 8]
```diff
@@ -987,16 +1011,49 @@ def __init__(
...
+                raise ValueError("`d` has to be provided in case of using a callable.")
+            if (indices is not None) and (len(indices) == 0):
+                raise ValueError("`indices` list is empty!")
+            indices = normalize_indices(indices=indices, d=d)
+            if (indices is not None) and (len(indices) > 0):
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py]: Thx for the suggestion, I fixed it.

File: botorch/models/transforms/input.py [hunk 9]
```diff
@@ -1017,15 +1074,88 @@ def transform(self, X: Tensor) -> Tensor:
         Returns:
             A `batch_shape x (q * n_f) x (d + d_f)`-dim tensor of appended inputs.
         """
+        if self._use_callable:
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply botorch/models/transforms/input.py]: Your are correct, I fixed it!

**Comment by jduerholt** [comment]
Regarding the `skip_expand`, I am not sure, that I undesrtand it 100%, which blocks me from writing the test case. For our use case, we do not need it currently. So for me it would be fine to not support it in the moment. If it should be supported, I have to look a bit deeper into it ...

**Review by saitcakmak — APPROVED** [review]
Thanks for the updates! This looks good to me. No need to worry about `skip_expand`, I can deal with that if needed later.
